src/HW_models/NRF_RTC.c

```diff
@@ -125,12 +126,67 @@ static void time_of_1_counter_wrap(int rtc, bs_time_t *us, uint16_t *frac) {
   *us = counter_to_time((uint64_t)RTC_COUNTER_MASK + 1, rtc, frac);
 }
 
+/**
+ * Return the time in microseconds when a given counter value (<rtc_counter>)
+ * will be reached next time for the RTC number (<rtc>).
+ * Note that if this counter value has just been reached, it will return the next match.
+ * That is, one full period later.
+ * @note To obtain a time delta, use @ref counter_to_time instead
+ */
+static bs_time_t get_time_of_next_counter_match(int rtc, uint32_t rtc_counter) {
+  uint16_t frac;
+  bs_time_t next_match_time = RTC_counter_startT[rtc]
+                              + counter_to_time(rtc_counter, rtc, &frac);
+
+  bs_time_t now = tm_get_hw_time();
+
+  if (next_match_time <= now) {
+    bs_time_t t_us;
+    uint16_t t_frac;
+
+    time_of_1_counter_wrap(rtc, &t_us, &t_frac);
+
+    do {
+      next_match_time += t_us;
+      frac += t_frac;
+      if (frac >= 0x200){
+        frac -= 0x200;
+        next_match_time +=1;
+      }
+    } while (next_match_time <= now);
+  }
+
+  return next_match_time;
+}
+
+/**
+ * Updates the time where the next tick event will be triggered
+ */
+static void update_next_match_events_tick(int rtc) {
+  /* When the tick event is enabled, a tick is generated at every increment
+   * of the timer. Here we calculate the time where the next tick occurs.
+   * To be able to know what the next counter value is, we first ensure
+   * that the current counter value is up to date.
+   */
+  nrf_rtc_update_COUNTER(rtc);
+  bs_time_t next_match_time = get_time_of_next_counter_match(rtc, (NRF_RTC_regs[rtc].COUNTER + 1) & RTC_COUNTER_MASK);
+  timers_tick[rtc] = next_match_time;
```

Comment:
There is 2 things here which are not functionally wrong, but are a bit ugly:
* When a tick interrupt or event is disabled, you leave behind its `timers_tick` with the old value. That won't trigger an spurious event up to other HW or IRQ with the checks you have, but it will trigger an spurious run of this HW model. It would be conceptually nicer if timers_tick[i] was "cleared" to TIME_NEVER in that case.
* `update_master_timer()` is now not the only place where the master timer (`Timer_RTC`) is updated, and where an update  is propagated up to the "time machine". (Spreading functionality around tends to make for more confusing/difficult to maintain code). I would be happier if:
* this function only updated `timers_tick[]`
* `update_master_timer()` would go thru `timers_tick[]` like it does for `cc_timers[][]`
* and you called `update_master_timer()` in the places where you added the direct call to `update_next_match_events_tick()`